crates/astria-conductor/tests/blackbox/firm_only.rs

```diff
@@ -377,3 +406,109 @@ async fn fetch_from_later_celestia_height() {
     );
 }
+
+#[tokio::test(flavor = "multi_thread", worker_threads = 1)]
+async fn exits_on_celestia_chain_id_mismatch() {
+    use astria_grpc_mock::{
+        matcher,
+        response as GrpcResponse,
+        Mock as GrpcMock,
+    };
+
+    // We have to create our own test conductor and perform mounts manually because `TestConductor`
+    // implements the `Drop` trait, which disallows us from taking ownership of its tasks and
+    // awaiting their completion.
+
+    let mock_grpc = MockGrpc::spawn().await;
+    let mock_http = wiremock::MockServer::start().await;
+
+    let config = Config {
+        celestia_node_http_url: mock_http.uri(),
+        execution_rpc_url: format!("http://{}", mock_grpc.local_addr),
+        sequencer_cometbft_url: mock_http.uri(),
+        sequencer_grpc_url: format!("http://{}", mock_grpc.local_addr),
+        execution_commit_level: CommitLevel::FirmOnly,
+        ..make_config()
+    };
+
+    let (metrics, _) = metrics::ConfigBuilder::new()
+        .set_global_recorder(false)
+        .build(&())
+        .unwrap();
+    let metrics = Box::leak(Box::new(metrics));
+
+    let conductor = {
+        let conductor = Conductor::new(config, metrics).unwrap();
+        conductor.spawn()
+    };
+
+    GrpcMock::for_rpc_given(
+        "get_genesis_info",
+        matcher::message_type::<GetGenesisInfoRequest>(),
+    )
+    .respond_with(GrpcResponse::constant_response(
+        genesis_info!(sequencer_genesis_block_height: 1,
+            celestia_block_variance: 10,),
+    ))
+    .expect(0..)
+    .mount(&mock_grpc.mock_server)
+    .await;
+
+    GrpcMock::for_rpc_given(
+        "get_commitment_state",
+        matcher::message_type::<GetCommitmentStateRequest>(),
+    )
+    .respond_with(GrpcResponse::constant_response(commitment_state!(firm: (
+            number: 1,
+            hash: [1; 64],
+            parent: [0; 64],
+        ),
+        soft: (
+            number: 1,
+            hash: [1; 64],
+            parent: [0; 64],
+        ),
+        base_celestia_height: 1,)))
+    .expect(0..)
+    .mount(&mock_grpc.mock_server)
+    .await;
+
+    Mock::given(body_partial_json(
+        json!({"jsonrpc": "2.0", "method": "header.NetworkHead"}),
+    ))
+    .and(header(
+        "authorization",
+        &*format!("Bearer {CELESTIA_BEARER_TOKEN}"),
+    ))
+    .respond_with(ResponseTemplate::new(200).set_body_json(json!({
+        "jsonrpc": "2.0",
+        "id": 0,
+        "result": celestia_network_head!(height: 1u32, chain_id: "bad_chain_id"),
+    })))
+    .expect(1..)
+    .mount(&mock_http)
+    .await;
+
+    let res = conductor.await;
+    match res {
+        Ok(Ok(())) => panic!("conductor should have exited with an error, no error received"),
+        Ok(Err(e)) => {
+            let mut source = e.source();
+            while source.is_some() {
+                let err = source.unwrap();
+                if let Some(CelestiaChainIdError::MismatchedCelestiaChainId {
```

Comment:
As above, don't assert on implementation details in blackbox tests. Just check behaviour.

Reply:
With checking the error source text itself, I've managed to work it such that the tests still check the error result without exposing any underlying types. It feels like the tests should probably ensure that the correct thing is causing the error, as opposed to succeeding for any kind of early exit, but I understand this is fragile so I can change if needed.